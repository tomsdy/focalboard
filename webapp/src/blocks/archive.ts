// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.
import {IBlock} from './block'

interface IArchiveHeader {
    version: number
    date: number
}

interface IArchiveLine {
    type: string,
    data: any,
}

// This schema allows the expansion of additional line types in the future
interface IBlockArchiveLine extends IArchiveLine {
    type: 'block',
    data: IBlock
}

class ArchiveUtils {
    static buildBlockArchive(blocks: readonly IBlock[]): string {
        const header: IArchiveHeader = {
            version: 1,
            date: Date.now(),
        }

        const headerString = JSON.stringify(header)
        let content = headerString + '\n'
        for (const block of blocks) {
            const line: IBlockArchiveLine = {
                type: 'block',
                data: block,
            }
            const lineString = JSON.stringify(line)
            content += lineString
            content += '\n'
        }

        return content
    }

    static parseBlockArchive(contents: string): IBlock[] {
        const blocks: IBlock[] = []
        const allLineStrings = contents.split('\n')
        if (allLineStrings.length >= 2) {
            const headerString = allLineStrings[0]
            const header = JSON.parse(headerString) as IArchiveHeader
            if (header.date && header.version >= 1) {
                const lineStrings = allLineStrings.slice(1)
                let lineNum = 2
                for (const lineString of lineStrings) {
                    if (!lineString) {
                        // Ignore empty lines, e.g. last line
                        continue
                    }
                    const line = JSON.parse(lineString) as IArchiveLine
                    if (!line || !line.type || !line.data) {
                        throw new Error(`ERROR parsing line ${lineNum}`)
                    }
                    switch (line.type) {
                    case 'block': {
                        const blockLine = line as IBlockArchiveLine
                        const block = blockLine.data
                        blocks.push(block)
                        break
                    }
                    }

                    lineNum += 1
                }
            } else {
                throw new Error('ERROR parsing header')
            }
        }

        return blocks
    }
}

export {IArchiveHeader, IArchiveLine, IBlockArchiveLine, ArchiveUtils}
