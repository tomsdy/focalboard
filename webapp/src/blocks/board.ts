// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.
import {Utils} from '../utils'

import {IBlock, MutableBlock} from './block'

type PropertyType = 'text' | 'number' | 'select' | 'multiSelect' | 'date' | 'person' | 'file' | 'checkbox' | 'url' | 'email' | 'phone' | 'createdTime' | 'createdBy' | 'updatedTime' | 'updatedBy'

interface IPropertyOption {
    readonly id: string
    readonly value: string
    readonly color: string
}

interface IMutablePropertyOption {
    id: string
    value: string
    color: string
}

// A template for card properties attached to a board
interface IPropertyTemplate {
    readonly id: string
    readonly name: string
    readonly type: PropertyType
    readonly options: IPropertyOption[]
}

interface IMutablePropertyTemplate extends IPropertyTemplate {
    id: string
    name: string
    type: PropertyType
    options: IMutablePropertyOption[]
}

interface Board extends IBlock {
    readonly icon: string
    readonly description: string
    readonly showDescription: boolean
    readonly isTemplate: boolean
    readonly cardProperties: readonly IPropertyTemplate[]
    duplicate(): MutableBoard
}

class MutableBoard extends MutableBlock implements Board {
    get icon(): string {
        return this.fields.icon as string
    }
    set icon(value: string) {
        this.fields.icon = value
    }

    get description(): string {
        return this.fields.description as string
    }
    set description(value: string) {
        this.fields.description = value
    }

    get showDescription(): boolean {
        return Boolean(this.fields.showDescription)
    }
    set showDescription(value: boolean) {
        this.fields.showDescription = value
    }

    get isTemplate(): boolean {
        return Boolean(this.fields.isTemplate)
    }
    set isTemplate(value: boolean) {
        this.fields.isTemplate = value
    }

    get cardProperties(): IMutablePropertyTemplate[] {
        return this.fields.cardProperties as IPropertyTemplate[]
    }
    set cardProperties(value: IMutablePropertyTemplate[]) {
        this.fields.cardProperties = value
    }

    constructor(block: any = {}) {
        super(block)
        this.type = 'board'

        this.icon = block.fields?.icon || ''
        this.description = block.fields?.description || ''
        if (block.fields?.cardProperties) {
            // Deep clone of card properties and their options
            this.cardProperties = block.fields.cardProperties.map((o: IPropertyTemplate) => {
                return {
                    id: o.id,
                    name: o.name,
                    type: o.type,
                    options: o.options ? o.options.map((option) => ({...option})) : [],
                }
            })
        } else {
            this.cardProperties = []
        }
    }

    duplicate(): MutableBoard {
        const board = new MutableBoard(this)
        board.id = Utils.createGuid()
        return board
    }
}

export {Board, MutableBoard, PropertyType, IPropertyOption, IPropertyTemplate}
