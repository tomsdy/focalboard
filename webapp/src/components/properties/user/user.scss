.UserProperty {
    margin-right: 20px;
    min-width: 120px;

    &.readonly {
        overflow: hidden;
        text-overflow: ellipsis;
        min-width: unset;
    }
}
