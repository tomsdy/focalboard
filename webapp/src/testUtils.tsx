// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.
import {IntlProvider} from 'react-intl'
import React from 'react'

export const wrapIntl = (children: any) => <IntlProvider locale='en'>{children}</IntlProvider>

export function mockDOM(): void {
    window.focus = jest.fn()
    document.createRange = () => {
        const range = new Range()

        range.getBoundingClientRect = jest.fn()

        range.getClientRects = () => {
            return {
                item: () => null,
                length: 0,
                [Symbol.iterator]: jest.fn(),
            }
        }

        return range
    }
}
