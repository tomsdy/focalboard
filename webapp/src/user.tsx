// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import React from 'react'

const UserContext = React.createContext(undefined as IUser|undefined)

const WorkspaceUsersContext = React.createContext({
    users: new Array<IUser>(),
    usersById: new Map<string, IUser>(),
})

interface IUser {
    id: string,
    username: string,
    email: string,
    props: Record<string, any>,
    createAt: number,
    updateAt: number,
}

type WorkspaceUsersContextData = {
    users: Array<IUser>
    usersById: Map<string, IUser>
}

export {IUser, UserContext, WorkspaceUsersContext, WorkspaceUsersContextData}
