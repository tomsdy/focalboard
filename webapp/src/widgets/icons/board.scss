@use './standardIcon.scss';

.focalboard-body .BoardIcon {
    @extend .StandardIcon;
    stroke-width: 8px;
}
