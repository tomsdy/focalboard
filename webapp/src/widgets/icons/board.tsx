// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import React from 'react'

import './board.scss'

export default function BoardIcon(): JSX.Element {
    return (
        <svg
            xmlns='http://www.w3.org/2000/svg'
            className='BoardIcon Icon'
            viewBox='0 0 100 100'
        >
            <rect
                x='10'
                y='10'
                width='80'
                height='80'
                rx='5'
                ry='5'
            />
            <polyline
                points='28,25 28,55'
                style={{strokeWidth: '15px'}}
            />
            <polyline
                points='50,25 50,70'
                style={{strokeWidth: '15px'}}
            />
            <polyline
                points='72,25 72,45'
                style={{strokeWidth: '15px'}}
            />
        </svg>
    )
}
