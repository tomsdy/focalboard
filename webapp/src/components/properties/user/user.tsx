// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import React, {useContext} from 'react'
import Select from 'react-select'

import {IUser, WorkspaceUsersContext, WorkspaceUsers} from '../../../user'

import './user.scss'
import {getSelectBaseStyle} from '../../../theme'

type Props = {
    value: string,
    readonly: boolean,
    onChange: (value: string) => void,
}

const UserProperty = (props: Props): JSX.Element => {
    const workspaceUsers = useContext<WorkspaceUsers>(WorkspaceUsersContext)

    if (props.readonly) {
        return (<div className='UserProperty octo-propertyvalue readonly'>{workspaceUsers.usersById.get(props.value)?.username || props.value}</div>)
    }

    return (
        <Select
            options={workspaceUsers.users}
            isSearchable={true}
            isClearable={true}
            backspaceRemovesValue={true}
            className={'UserProperty'}
            styles={getSelectBaseStyle()}
            getOptionLabel={(o: IUser) => o.username}
            getOptionValue={(a: IUser) => a.id}
            value={workspaceUsers.usersById.get(props.value) || null}
            onChange={(item, action) => {
                if (action.action === 'select-option') {
                    props.onChange(item?.id || '')
                } else if (action.action === 'clear') {
                    props.onChange('')
                }
            }}
        />
    )
}

export default UserProperty
