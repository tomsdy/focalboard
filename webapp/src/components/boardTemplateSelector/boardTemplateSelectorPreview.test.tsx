// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.
import {render, waitFor} from '@testing-library/react'
import React from 'react'
import {MockStoreEnhanced} from 'redux-mock-store'

import {Provider as ReduxProvider} from 'react-redux'

import {UserWorkspace} from '../../user'
import {IPropertyTemplate} from '../../blocks/board'
import {mockDOM, mockStateStore, wrapDNDIntl} from '../../testUtils'

import BoardTemplateSelectorPreview from './boardTemplateSelectorPreview'

jest.mock('react-router-dom', () => {
    const originalModule = jest.requireActual('react-router-dom')

    return {
        ...originalModule,
        useRouteMatch: jest.fn(() => {
            return {url: '/'}
        }),
    }
})

const groupProperty: IPropertyTemplate = {
    id: 'group-prop-id',
    name: 'name',
    type: 'text',
    options: [
        {
            color: 'propColorOrange',
            id: 'property_value_id_1',
            value: 'Q1',
        },
        {
            color: 'propColorBlue',
            id: 'property_value_id_2',
            value: 'Q2',
        },
    ],
}

jest.mock('../../octoClient', () => {
    return {
        getSubtree: jest.fn(() => Promise.resolve([
            {
                id: '1',
                workspaceId: 'workspace',
                title: 'Template',
                type: 'board',
                fields: {
                    icon: '🚴🏻‍♂️',
                    cardProperties: [groupProperty],
                    dateDisplayPropertyId: 'id-5',
                },
            },
            {
                id: '2',
                workspaceId: 'workspace',
                title: 'View',
                type: 'view',
                fields: {
                    groupById: 'group-prop-id',
                    viewType: 'board',
                    visibleOptionIds: ['group-prop-id'],
                    hiddenOptionIds: [],
                    visiblePropertyIds: ['group-prop-id'],
                    sortOptions: [],
                    kanbanCalculations: {},
                },
            },
            {
                id: '3',
                workspaceId: 'workspace',
                title: 'Card',
                type: 'card',
                fields: {
                    icon: '🚴🏻‍♂️',
                    properties: {
                        'group-prop-id': 'test',
                    },
                },
            },
        ])),
    }
})
jest.mock('../../utils')
jest.mock('../../mutator')

describe('components/boardTemplateSelector/boardTemplateSelectorPreview', () => {
    const workspace1: UserWorkspace = {
        id: 'workspace_1',
        title: 'Workspace 1',
        boardCount: 1,
    }
    const template1Title = 'Template 1'
    const globalTemplateTitle = 'Template Global'
    const boardTitle = 'Board 1'
    let store:MockStoreEnhanced<unknown, unknown>
    beforeAll(mockDOM)
    beforeEach(() => {
        jest.clearAllMocks()
        const state = {
            searchText: {value: ''},
            users: {me: {id: 'user-id'}},
            cards: {templates: []},
            views: {views: []},
            contents: {contents: []},
            comments: {comments: []},
            workspace: {
                userWorkspaces: new Array<UserWorkspace>(workspace1),
                current: workspace1,
            },
            boards: {
                boards: [
                    {
                        id: '2',
                        title: boardTitle,
                        workspaceId: workspace1.id,
                        fields: {
                            icon: '🚴🏻‍♂️',
                            cardProperties: [groupProperty],
                            dateDisplayPropertyId: 'id-6',
                        },
                    },
                ],
                templates: [
                    {
                        id: '1',
                        workspaceId: workspace1.id,
                        title: template1Title,
                        fields: {
                            icon: '🚴🏻‍♂️',
                            cardProperties: [groupProperty],
                            dateDisplayPropertyId: 'id-5',
                        },
                    },
                ],
                cards: [],
                views: [],
            },
            globalTemplates: {
                value: [{
                    id: 'global-1',
                    title: globalTemplateTitle,
                    workspaceId: '0',
                    fields: {
                        icon: '🚴🏻‍♂️',
                        cardProperties: [
                            {id: 'global-id-5'},
                        ],
                        dateDisplayPropertyId: 'global-id-5',
                    },
                }],
            },
        }
        store = mockStateStore([], state)
    })

    test('should match snapshot', async () => {
        const {container} = render(wrapDNDIntl(
            <ReduxProvider store={store}>
                <BoardTemplateSelectorPreview activeTemplate={(store.getState() as any).boards.templates[0]}/>
            </ReduxProvider>
            ,
        ))
        await waitFor(() => expect(container.querySelector('.top-head')).not.toBeNull())
        expect(container).toMatchSnapshot()
    })
    test('should be null without activeTemplate', () => {
        const {container} = render(wrapDNDIntl(
            <ReduxProvider store={store}>
                <BoardTemplateSelectorPreview activeTemplate={null}/>
            </ReduxProvider>
            ,
        ))
        expect(container).toMatchSnapshot()
    })
})
