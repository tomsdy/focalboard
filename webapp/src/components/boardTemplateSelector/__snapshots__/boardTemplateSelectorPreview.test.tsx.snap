// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/boardTemplateSelector/boardTemplateSelectorPreview should be null without activeTemplate 1`] = `<div />`;

exports[`components/boardTemplateSelector/boardTemplateSelectorPreview should match snapshot 1`] = `
<div>
  <div
    class="BoardTemplateSelectorPreview"
  >
    <div
      class="prevent-click"
    />
    <div
      class="top-head"
    >
      <div
        class="ViewTitle"
      >
        <div
          class="add-buttons add-visible"
        />
        <div
          class="title"
        >
          <div
            class="BlockIconSelector"
          >
            <div
              aria-label="menuwrapper"
              class="MenuWrapper"
              role="button"
            >
              <div
                class="octo-icon size-m"
              >
                <span>
                  üö¥üèª‚Äç‚ôÇÔ∏è
                </span>
              </div>
            </div>
          </div>
          <input
            class="Editable readonly title"
            placeholder="Untitled board"
            readonly=""
            spellcheck="true"
            title="Template 1"
            value="Template 1"
          />
        </div>
      </div>
      <div
        class="ViewHeader"
      >
        <input
          class="Editable undefined"
          placeholder="Untitled View"
          spellcheck="true"
          title="View"
          value="View"
        />
        <div
          aria-label="View menu"
          class="MenuWrapper"
          role="button"
        >
          <button
            type="button"
          >
            <i
              class="CompassIcon icon-chevron-down DropdownIcon"
            />
          </button>
        </div>
        <div
          class="octo-spacer"
        />
        <div
          aria-label="Properties menu"
          class="MenuWrapper"
          role="button"
        >
          <button
            type="button"
          >
            <span>
              Properties
            </span>
          </button>
        </div>
        <div
          aria-label="menuwrapper"
          class="MenuWrapper"
          role="button"
        >
          <button
            type="button"
          >
            <span>
              Group by: 
              <span
                id="groupByLabel"
              >
                name
              </span>
            </span>
          </button>
        </div>
        <div
          class="ModalWrapper"
        >
          <button
            type="button"
          >
            <span>
              Filter
            </span>
          </button>
        </div>
        <div
          aria-label="menuwrapper"
          class="MenuWrapper"
          role="button"
        >
          <button
            type="button"
          >
            <span>
              Sort
            </span>
          </button>
        </div>
        <button
          type="button"
        >
          <span>
            Search
          </span>
        </button>
        <div
          class="ModalWrapper"
        >
          <div
            aria-label="View header menu"
            class="MenuWrapper"
            role="button"
          >
            <button
              type="button"
            >
              <i
                class="CompassIcon icon-dots-horizontal OptionsIcon"
              />
            </button>
          </div>
        </div>
        <div
          class="ButtonWithMenu"
        >
          <div
            class="button-text"
          >
            New
          </div>
          <div
            aria-label="menuwrapper"
            class="MenuWrapper"
            role="button"
          >
            <div
              class="button-dropdown"
            >
              <i
                class="CompassIcon icon-chevron-down DropdownIcon"
              />
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="Kanban"
    >
      <div
        class="octo-board-header"
        id="mainBoardHeader"
      >
        <div
          class="octo-board-header-cell KanbanColumnHeader"
          draggable="true"
          style="opacity: 1;"
        >
          <span
            class="Label empty "
            title="Items with an empty name property will go here. This column cannot be removed."
          >
            No name
          </span>
          
          <div
            class="KanbanCalculation"
          >
            <button
              title="1"
              type="button"
            >
              <span>
                1
              </span>
            </button>
          </div>
          <div
            class="octo-spacer"
          />
          <div
            aria-label="menuwrapper"
            class="MenuWrapper"
            role="button"
          >
            <button
              type="button"
            >
              <i
                class="CompassIcon icon-dots-horizontal OptionsIcon"
              />
            </button>
          </div>
          <button
            type="button"
          >
            <i
              class="CompassIcon icon-plus AddIcon"
            />
          </button>
        </div>
        <div
          class="octo-board-header-cell KanbanColumnHeader"
          draggable="true"
          style="opacity: 1;"
        >
          <span
            class="Label propColorOrange "
          >
            <input
              class="Editable undefined"
              placeholder="New Select"
              spellcheck="true"
              title="Q1"
              value="Q1"
            />
          </span>
          <div
            class="KanbanCalculation"
          >
            <button
              title="0"
              type="button"
            >
              <span>
                0
              </span>
            </button>
          </div>
          <div
            class="octo-spacer"
          />
          <div
            aria-label="menuwrapper"
            class="MenuWrapper"
            role="button"
          >
            <button
              type="button"
            >
              <i
                class="CompassIcon icon-dots-horizontal OptionsIcon"
              />
            </button>
          </div>
          <button
            type="button"
          >
            <i
              class="CompassIcon icon-plus AddIcon"
            />
          </button>
        </div>
        <div
          class="octo-board-header-cell KanbanColumnHeader"
          draggable="true"
          style="opacity: 1;"
        >
          <span
            class="Label propColorBlue "
          >
            <input
              class="Editable undefined"
              placeholder="New Select"
              spellcheck="true"
              title="Q2"
              value="Q2"
            />
          </span>
          <div
            class="KanbanCalculation"
          >
            <button
              title="0"
              type="button"
            >
              <span>
                0
              </span>
            </button>
          </div>
          <div
            class="octo-spacer"
          />
          <div
            aria-label="menuwrapper"
            class="MenuWrapper"
            role="button"
          >
            <button
              type="button"
            >
              <i
                class="CompassIcon icon-dots-horizontal OptionsIcon"
              />
            </button>
          </div>
          <button
            type="button"
          >
            <i
              class="CompassIcon icon-plus AddIcon"
            />
          </button>
        </div>
        <div
          class="octo-board-header-cell narrow"
        >
          <button
            type="button"
          >
            <span>
              + Add a group
            </span>
          </button>
        </div>
      </div>
      <div
        class="octo-board-body"
        id="mainBoardBody"
      >
        <div
          class="octo-board-column"
        >
          <div
            class="KanbanCard"
            draggable="true"
            style="opacity: 1;"
          >
            <div
              aria-label="menuwrapper"
              class="MenuWrapper optionsMenu"
              role="button"
            >
              <button
                type="button"
              >
                <i
                  class="CompassIcon icon-dots-horizontal OptionsIcon"
                />
              </button>
            </div>
            <div
              class="octo-icontitle"
            >
              <div
                class="octo-icon"
              >
                üö¥üèª‚Äç‚ôÇÔ∏è
              </div>
              <div
                class="octo-titletext"
              >
                Card
              </div>
            </div>
            <div
              class="octo-tooltip tooltip-top"
              data-tooltip="name"
            >
              <div
                class="octo-propertyvalue octo-propertyvalue--readonly"
              >
                test
              </div>
            </div>
          </div>
          <button
            type="button"
          >
            <span>
              + New
            </span>
          </button>
        </div>
        <div
          class="octo-board-column"
        >
          <button
            type="button"
          >
            <span>
              + New
            </span>
          </button>
        </div>
        <div
          class="octo-board-column"
        >
          <button
            type="button"
          >
            <span>
              + New
            </span>
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;
